src/test/scala/components/AsyncMemoryUnitTest.scala

```diff
@@ -0,0 +1,194 @@
+// Unit tests for the async memory objects.
+
+package dinocpu
+
+import chisel3.iotesters
+import chisel3.iotesters.{ChiselFlatSpec, Driver, PeekPokeTester}
+
+// To generate random latencies
+import scala.util.Random
+
+
+// Test instruction port reading a zeroed memory file
+class AsyncMemoryUnitTester$IMemZero(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // Expect 0's on the instruction port
+  for (i <- 0 to size/4 - 1) {
+    poke(m.io.imem.address, i*4)
+    poke(m.io.imem.ready, 1)
+
+    step(latency)
+    
+    expect(m.io.imem.valid, 1)
+    expect(m.io.imem.instruction, 0)
+  }
+}
+
+// Test instruction port reading an ascending memory file
+class AsyncMemoryUnitTester$IMemRead(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // Expect ascending bytes on instruction port
+  for (i <- 0 to size/4 - 1) {
+    poke(m.io.imem.address, i*4)
+    poke(m.io.imem.ready, 1)
+    
+    step(latency)
+    
+    expect(m.io.imem.valid, 1)
+    expect(m.io.imem.instruction, i)
+  }
+}
+
+// Test data port writes and instruction port reads
+class AsyncMemoryUnitTester$IMemWrite(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // write ascending data to memory 
+  for (i <- 0 to size/8 - 1) {
+    poke(m.io.dmem.address, i*4)
+    poke(m.io.dmem.memwrite, 1)
+    poke(m.io.dmem.maskmode, 2)
+    poke(m.io.dmem.sext, 0)
+    poke(m.io.dmem.writedata, i+100)
+
+    step(latency)
+
+    expect(m.io.dmem.valid, 1)
+  }
+
+  poke (m.io.dmem.memwrite, 0)
+
+  // expect ascending bytes on instruction port
+  for (i <- 0 to size/4 - 1) {
+    poke(m.io.imem.address, i*4)
+    poke(m.io.imem.ready, 1)
+    
+    step(latency)
+   
+    expect(m.io.imem.valid, 1)
+    if (i < size/8) {
+      expect(m.io.imem.instruction, i+100)
+    } else {
+      expect(m.io.imem.instruction, i)
+    }
+  }
+}
+
+// Test data port reading a zeroed memory file
+class AsyncMemoryUnitTester$DMemZero(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
```

Comment:
What does a `$` in a type name mean? Anything special? I'm not familiar with this scala syntax.

Reply:
In Java you can use dollar symbols in class names as characters, and it's primarily used to generate class files of an inner class inside another class (like Parent$Child.class). It doesn't really mean anything special in Java or Scala, but I just used it to represent a similar concept of there being a "DMemZero" sub-test inside the "AsyncMemoryUnitTester" that tests the data memory for zeroes. 

The other alternative I had was to attempt to follow the existing naming scheme in MemoryUnitTest.scala, which resulted in long names like "AsyncMemoryUnitDMemWriteTester" that are hard to parse at first glance. I do agree that the use of the dollar symbol is strange though, so I considered using underscores, but combining camel case and snake case just looked awkward to me.

Reply:
Gotcha. I understand now. This is just something I hadn't seen before. :)